### PR TITLE
More list properties about `catMaybes/mapMaybe`

### PR DESCRIPTION
**Pull request by omelkonian**
- Follow-up to PR #2361
- Ported from my [formal-prelude](https://github.com/omelkonian/formal-prelude/tree/c10fe94876a7378088f2e4cf68d9b18858dcc256)

### CODE REVIEW

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [inline src/Data/List/Properties.agda:86]: Any strong reason to re-order the file this way?

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [inline src/Data/List/Properties.agda:727]: Personally I think it's clearer to not do 2-level pattern matching but to delegate the job to `maybe`, i.e.
```agda
length-catMaybes (x  ∷ xs) = let l = length-catMaybes xs) in maybe (\ _ → s≤s l) (m≤n⇒m≤1+n l) x
```
which more cleanly separates the list-based reasoning from the maybe-based reasoning.

(In my own code, I'd got further and use higher-order instead of `let` but I've been convinced that this is a step too far for stdlib)

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [inline src/Data/List/Properties.agda:803]: why not pattern-match ? This doesn't match the rest of the style.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (JacquesCarette) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:385]: `maybe` would be better than `with` here.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (JacquesCarette) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:389]: whereas this indicates a missing combinator in `Data.Maybe.Base`, a more general version of `zipWith`.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda:727]: Also: `where` instead of `let` cf. your own #2369 ?

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:54]: I don't see this anywhere in `docs/style-guide.md` btw. Maybe add an example like the following to make it clear:
```agda
private 
  abstract 
    f : ...
    f = ...
```

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:803]: I prefer less ink on the page, especially if it the added information is completely useless/redundant (in this case repeating the name `mapMaybeIsInj₁∘mapInj₁`).

It gets even worse when you have to repeat arguments as well; maybe the case of `Any-catMaybes⁺` convinces you...

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:727]: >  Personally I think it's clearer to not do 2-level pattern matching but to delegate the job to maybe

Your suggested change doesn't typecheck though: using eliminators in dependent situations often requires that you explicitly provide the 'motive' (in this case generalising over `mx`). Indeed that's what people used to do before proper dependent pattern matching came along:
```agda
length-catMaybes : ∀ xs → length (catMaybes {A = A} xs) ≤ length xs
length-catMaybes []        = ≤-refl
length-catMaybes (mx ∷ xs) = let IH = length-catMaybes xs in
  maybe {B = λ mx → length (catMaybes (mx ∷ xs)) ≤ length (mx ∷ xs)}
        (λ _ → s≤s IH)
        (m≤n⇒m≤1+n IH)
        mx
```
It gets even worse in more complicated scenarios like `catMaybes-↭` where you need to introduce more variables, etc.

A quick `grep` through the codebase reveals no other explicit motives for `maybe`: are you sure you want this change (and most importantly, maintaining this code)?

Btw, `length-catMaybes` was already there :rofl:

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (omelkonian) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:385]: See comment above.

Comment (omelkonian) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:385]: For reference, this is what you would need to write:
```agda
catMaybes-↭ {xs = .mx ∷ xs} {.mx ∷ ys} (prep mx xs↭)
  = maybe {B = λ mx → catMaybes (mx ∷ xs) ↭ catMaybes (mx ∷ ys)}
          (λ x → prep x $ catMaybes-↭ xs↭)
          (catMaybes-↭ xs↭)
          mx
```
Notice the necessity to introduce variables for the tails of the two lists.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (omelkonian) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:389]: I don't get what you're suggesting here: what does `zipWith` have to do with this proof? 

Given your previous comments, you probably want an eliminator for two maybes simultaneously?
```agda
maybe² : ∀ {A : Set a} {B : Maybe A → Maybe A → Set b}
  → (∀ x y → B (just x) (just y))
  → (∀ x → B (just x) nothing)
  → (∀ y → B nothing (just y))
  → B nothing nothing
  → (x y : Maybe A)
  → B x y
maybe² jj jn nj nn = λ where
  (just x) (just y) → jj x y
  (just x) nothing  → jn x
  nothing  (just y) → nj y
  nothing  nothing  → nn
```
Ofc, the motive will again be necessary:
```agda
catMaybes-↭ {xs = .mx ∷ .my ∷ xs} {.my ∷ .mx ∷ ys} (swap mx my xs↭)  =
    let IH = catMaybes-↭ xs↭ in
    maybe² {B = λ mx my → catMaybes (mx ∷ my ∷ xs) ↭ catMaybes (my ∷ mx ∷ ys)}
           (λ x y → swap x y IH)
           (λ x → prep x IH)
           (λ y → prep y IH)
           IH
           mx my
```

**Comment by omelkonian** [comment]
@JacquesCarette To sum up, let's agree that there are both pros and cons to using old-school eliminators in contrast to immediately using (dependent) pattern matching:
- *PRO:* reusability! e.g. if we have some lemmas about `maybe` we could reuse them for deriving other proofs
- *CONS:*
   + type inference (e.g. explicitly giving the motives in the examples above)
   + someone wanting to prove something about these definitions can immediately see what they have to do case analysis on, while with `maybe` that have to perform one additional mental step
   
Given that there is only one lemma about `maybe` :disappointed: (`maybe-map`) and that the inductive hypotheses in the proofs above can easily be shared using `with IH ← ...`, I have not followed your suggestion to use eliminators and stuck with matching on the maybes and so forth.

Feel free to push another commit on top of my branch to convert to eliminator-style proofs though...

**Comment by JacquesCarette** [comment]
@omelkonian I'll comment on various of the details a bit later, but long story short: you make a very convincing case. I was overly optimistic.

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda]: or, indeed, with a 
```agda
private
  variable
    x y z : A
```
block...

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:759]: Why here rather than added after L150 in the original?

Comment (jamesmckinna) [reply src/Data/List/Properties.agda:759]: And for the sake of juxtaposition, I'd swap L758 and L759, as well as suppressing `x` and `xs`:
```suggestion
Any-catMaybes⁺ : ∀ {P : Pred A ℓ} {xs : List (Maybe A)}
  → Any (MAny P) xs → Any P (catMaybes xs)
Any-catMaybes⁺ {xs = nothing ∷ _} (there x∈)       = Any-catMaybes⁺ x∈
Any-catMaybes⁺ {xs = just _  ∷ _} (there x∈)       = there $ Any-catMaybes⁺ x∈
Any-catMaybes⁺ {xs = just _  ∷ _} (here (just px)) = here px
```

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:508]: Are these not (also) provable for the `Setoid` version?
If so, please add! (for me, that relation is conceptually prior to the `Propositional` one cf. #1761

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:743]: Notwithstanding your discussion with @JacquesCarette , I fail to see how these proofs improve on the original three-liners of L126-L149. What am I not grasping here? Esp. given your noted/notable aversion to `with`-based programming...

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:753]: Ditto.

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Properties.agda:759]: And here, the proof format follow the original style!?

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:37]: May not be in `style-guide`, but the additional blankline+indent  is customary everywhere else in `stdlib`...

**Review by jamesmckinna — CHANGES_REQUESTED** [review]
Great additions, but need to fix up layout and some of the `with`-based proofs in favou rof the (simpler!) originals, I think.

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Properties.agda]: #2390 at least documents that lacuna

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Properties.agda]: I was definitely overly optimistic regarding `maybe` being able to infer the motive. As it can't, then this 3-line proof is indeed the better one.

File: src/Data/List/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Properties.agda]: While I greatly sympathize with this 'less ink on the page', I have had to adjust my own style when contributing to `stdlib` where I think a certain amount of uniformity of style is better.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:386]: Right, that's worse.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (JacquesCarette) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:392]: I meant exactly this `maybe²` which can legitimately also be called `zipWith` for 2 `Maybe`s.

Having said that, clearly it wouldn't improve things (here). There might be a more-dependent version that would be able to deal with this, but that's out-of-scope for this PR.

**Comment by JacquesCarette** [comment]
As I agree with @jamesmckinna 's comments, I won't approve this PR (yet), even though I'm not happy with the parts that I had commented on myself.

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:743]: I used irrefutable `with` for reusability: the induction hypothesis is now only written once and shared across the branches of the case analysis. (the new feature for [left-hand side lets via the `using` keyword](https://github.com/agda/agda/pull/7078) would be even better, but this will have to wait until the next release)

Re. my `with`-scepticism, it mostly refers to basic definitions like `mapMaybe` and not so frequently when you prove properties about them. Especially when a use of `with` is unnecessary, that is you do not really need *dependent* pattern matching; here you actual need the dependency. All at the same time considering the reduction behaviour which might now favour the use of `with`, as mentioned in [@gallais' comment](https://github.com/agda/agda-stdlib/issues/1937#issuecomment-1483389966) in the original discussion about this.

 **However**, just to contradict my own argument above, when one gets into the business of proving meta-properties about these properties (e.g. you view proofs of membership [constructively as indices](https://omelkonian.github.io/formal-prelude/Prelude.Lists.Membership.html) --- ironically I also needed that for [permutations](https://omelkonian.github.io/formal-prelude/Prelude.Lists.PermutationsMeta.html) --- also c.f. notes about "meta-properties" in my [thesis](https://omelkonian.github.io/data/phd-thesis.pdf)), there is no distinction between "base" definitions and "properties". Having said that, the standard library does not include any of these to my knowledge, so we can just ignore these ultra-rare cases for the sake of practicality.

Therefore, I still think the current form sits nicely between the eliminator-style approach requested by Jacques and the original fully-direct non-reusing pattern matching.

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:759]: > Why here rather than added after L150 in the original?

Grouped the properties of `catMaybe` with the ones for `mapMaybe`, as it seemed more intuitive.
Also, lemma `catMaybes-concatMap` mentions `concatMap` which comes just before in the file.

Comment (omelkonian) [reply src/Data/List/Properties.agda:759]: Yep, here there is no shared expression to reuse, therefore I classify this case under 'unnecessary `with`s'.

Comment (JacquesCarette) [reply src/Data/List/Properties.agda:759]: I definitely prefer the 3-liner over this!

Comment (jamesmckinna) [reply src/Data/List/Properties.agda:759]: Well, perhaps there is no 'right' order for module contents, so the rule-of-thumb ends up being: try, where possible, to follow in `Properties` the order in `Base`... but even that seems hard to stick to.

File: CHANGELOG.md
Comment (omelkonian) [reply CHANGELOG.md:508]: Indeed they are, but require another module/file that is more general than `Data.List.Relation.Binary.Permutation.Setoid.Properties`, i.e. without a fixed module parameter for setoid equality on `A`. 

I added this as `Data.List.Relation.Binary.Permutation.Setoid.Properties.Maybe`.

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda]: I'm sympathetic to the argument based on sharing, but where it falls down for me is in the simple-minded LOC here... and some kind of trade-off in cognitive load that follows from that... sharing a recursive call both affords, but also *enforces*, a particular understanding of the call structure which is quite distinct from the one-clause-one-call-structure style. You might argue it's 'richer', but you also pay a higher price for it!

My instinct for `stdlib` goes towards a more 'middlebrow' approach... with a correspondingly lower bar to entry for newcomers?

Comment (jamesmckinna) [reply src/Data/List/Properties.agda]: As for 'meta' properties, I find increasingly the need for proofs of 'higher-dimensional' lifting lemmas (eg in [`Data.List.Relation.Binary.BagAndSetEquality`](https://github.com/agda/agda-stdlib/blob/c5255d0c28ca39b3233f4ade7c21a0a96b7a5b69/src/Data/List/Relation/Binary/BagAndSetEquality.agda#L579-L592)...

File: CHANGELOG.md
Comment (jamesmckinna) [reply CHANGELOG.md:508]: Ugh! yes... a wrinkle that recurs through the library. 
But can't such proofs follow eg the model of `Data.List.Relation.Binary.Permutation.Setoid.Properties.map`?

Comment (omelkonian) [reply CHANGELOG.md:508]: Unfortunately not: after flatting the maybes, you need to instantiate `catMaybes-↭` with type `B` instead of the (global) type `A`.

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda]: There are actually already occurrences of the `with ih ← ...` pattern already in the codebase + one could say that this is more intuitive to beginners than simultaneous deep pattern matching maybe?

File: CHANGELOG.md
Comment (jamesmckinna) [reply CHANGELOG.md:508]: Grrr. OK.

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties/Maybe.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties/Maybe.agda:71]: This `open` not required!

File: src/Data/List/Properties.agda
Comment (jamesmckinna) [reply src/Data/List/Properties.agda]: Indeed! ... I'm responsible for them... ;-)

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda
Comment (jamesmckinna) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:389]: In light of the `Setoid.Properties.Maybe` versions below, is there any longer any need to expose the forced implicit bindings for `xs` and `ys`? I'd reckon probably not...?

**Comment by jamesmckinna** [comment]
Thanks for the continued updates, but I won't re-review until you signal it's time to do so though!

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda
Comment (omelkonian) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:147]: Note that this is a side-fix: previously the function's co-domain (equivalence) was unnecessarily restricted to live at the same level as the domain (equivalence).

**Comment by omelkonian** [comment]
@jamesmckinna ready for review

File: CHANGELOG.md
Comment (andreasabel) [inline CHANGELOG.md:381]: Linebreak in awkward place here (break at "->" instead).

File: src/Data/List/Properties.agda
Comment (andreasabel) [inline src/Data/List/Properties.agda:803]: Do we have more general versions of this?  Looks like a property of a retract (inj1) aka embedding-projection pair.
So if `proj . emb = just`, then `mapMaybe proj . map emb = id`.

**Review by andreasabel — COMMENTED** [review]
Good stuff.

File: CHANGELOG.md
Comment (omelkonian) [reply CHANGELOG.md:381]: Not my change, but fixed that since you spotted it :)

File: src/Data/List/Properties.agda
Comment (omelkonian) [reply src/Data/List/Properties.agda:803]: You're absolutely right, now generalised.

**Review by andreasabel — APPROVED** [review]
Thanks for the update!

**Comment by omelkonian** [comment]
@JacquesCarette @jamesmckinna Ready to merge?

**Review by JacquesCarette — APPROVED** [review]
Yes from me.

File: CHANGELOG.md
Comment (jamesmckinna) [reply CHANGELOG.md:509]: See also #2397 this may require downstream refactoring...

**Review by jamesmckinna — COMMENTED** [review]
I don't think the new `with`-based code (and the accompanying code motion) is an improvement over the existing pattern-matching based definitions, but you have two approvals from others, so I'm not going to stand in the way.

**Comment by omelkonian** [comment]
> I don't think the new `with`-based code (and the accompanying code motion) is an improvement over the existing pattern-matching based definitions, but you have two approvals from others, so I'm not going to stand in the way.

Isn't the reuse of the inductive hypothesis (`ih`) clearly an improvement?

**Comment by jamesmckinna** [comment]
> > I don't think the new `with`-based code (and the accompanying code motion) is an improvement over the existing pattern-matching based definitions, but you have two approvals from others, so I'm not going to stand in the way.
> 
> Isn't the reuse of the inductive hypothesis (`ih`) clearly an improvement?

In this case, I (personally) actually think no, it's not, at least not enough to justify the rewriting of existing code. Funnily enough, I think that both you and @JacquesCarette have convinced me, in different ways, and from different starting points, to view such code style with... scepticism for the time being. And one reason I dismissed my review was to avoid going back over this ground...so: I stand aside, as previously.

**Comment by andreasabel** [comment]
Mmh, I tend to agree with James @jamesmckinna that `with` should be avoided where possible.  The reason is that it internally generates a helper function which Agda might sometimes leak in printing.  Also, it bears some hazards in proving things about the parent function as the name of the auxiliary function is not available to the user.
So, I think the general thrust should be to get rid of `with` where possible rather than introducing more `with`s.

**Comment by omelkonian** [comment]
> While I very much agree with @andreasabel 's general stance, especially for any definition that computes, I've gotten more relaxed about "pure proofs".

See my comment above on "meta-properties", which nullifies the distinction between programs and "pure proofs".

> Really what's we're sorely missing (and this is, I believe, what @omelkonian was using it for) is a proper pattern-matching `let` that doesn't unfold eagerly. This is a real wart in the current Agda language (along with the same issue wrt module parameters).

Isn't the upcoming `using` syntax that I mentioned above already what you wish for (although I'm not sure about its unfolding behaviour, but the `opaque` feature should cover that).

**Comment by JacquesCarette** [comment]
Yes, meta-properties, being proofs-about-proofs, do blur things. I've hated doing such proofs so much (in MLTT) that I block them out; when done 'properly', a la category theory where one adds higher-level cells, then these become sane again. `subst` hell is really the worst hell of them all.

I don't understand (yet!) enough of the details of `using` to know if it really going to be 'the' feature needed. I suspect that as long as there isn't some kind of `let`-like construct in the internal AST, it won't be, it'll merely be pleasant syntactic sugar.

**Comment by andreasabel** [comment]
@omelkonian: Thanks, Orestis, for your patience and willingness to revert changes!

**Comment by jamesmckinna** [comment]
> @omelkonian: Thanks, Orestis, for your patience and willingness to revert changes! (`with` is gone again)

Seconded!

**Pull request merged**
